test/badges.test.ts

```diff
@@ -228,48 +248,91 @@ describe('Merkle minting', () => {
     )
     await createSpec(badgesProxy, specUri, raftTokenId, issuer)
 
-    // ===== part 1: the issuer whitelists a bunch of addresses and saves signature + array of addresses to the database
     const whitelistAddresses = [claimant.address, '0x1', '0x2', '0x3']
 
-    // map over addresses to get leaf nodes
     const leafNodes = whitelistAddresses.map(addr => keccak256(addr))
 
-    // derive merkle tree
     const merkleTree = new MerkleTree(leafNodes, keccak256, { sortPairs: true })
 
-    // derive merkle root
     const merkleRoot = merkleTree.getRoot()
 
-    // embed merkle root into signature
     merkleTypedData.value.root = merkleRoot
 
-    // create signature
     const { compact } = await getSignature(
       merkleTypedData.domain,
       merkleTypedData.types,
       merkleTypedData.value,
       issuer
     )
-    // here we would store signature and `whitelistAddresses` in the db
-
-    // ===== part 2: simulate the claimant minting a badge
-    // claimant loads the webpage database returns signature and whitelistAddresses
 
-    // given the claimant's address, derive a merkleProof
     const merkleProof = merkleTree.getHexProof(leafNodes[0])
     await expect(
       badgesProxy
         .connect(claimant)
         .merkleTake(issuer.address, specUri, compact, merkleRoot, merkleProof)
     )
-    // pass merkleRoot and merkleProof to the badges contract
-    // badges contract will check to see if the proof is valid
 
-    // ===== part 3: check that the claimant has the badge
     expect(await badgesProxy.balanceOf(claimant.address)).equal(1)
   })
 
-  // todo: reject when someone who has already claimed tries to claim again
+  it.only('Should allow minting when an address is allowlisted on a merkle tree and both issuer and recipient provide signatures', async () => {
+    const {
+      badgesProxy,
+      raftProxy,
+      merkleTypedData,
+      issuer,
+      claimant,
+      owner,
+      requestTypedData,
+    } = deployed
+
+    const { raftTokenId } = await mintRaftToken(
+      raftProxy,
+      issuer.address,
+      specUri,
+      owner
+    )
+    await createSpec(badgesProxy, specUri, raftTokenId, issuer)
+
+    const whitelistAddresses = [claimant.address, '0x1', '0x2', '0x3']
```

Comment:
Can this block of code be refactored? If it's a big refactor, let's create another ticket but it looks to me that this is a generic block of code that is reused in other merkle tests too.

Reply:
yep, i'll refactor